### PR TITLE
Switch XContentBuilder from BytesStreamOutput to ByteArrayOutputStream

### PR DESCRIPTION
**Pull request by dakrone**
This switches the underlying byte output representation used by default in
`XContentBuilder` from `BytesStreamOutput` to a `ByteArrayOutputStream` (an
`OutputStream` can still be specified manually)

This is groundwork to allow us to decouple `XContent*` from the rest of the ES
core code so that it may be factored into a separate jar.

Since `BytesStreamOutput` was not using the recycling instance of `BigArrays`,
this should not affect the circuit breaking capabilities elsewhere in the
system.

Relates to #28504

### CODE REVIEW

**Review by rjernst — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 1]
```diff
@@ -1036,7 +1037,11 @@ public XContentGenerator generator() {
 
     public BytesReference bytes() {
         close();
-        return ((BytesStream) bos).bytes();
+        if (bos instanceof ByteArrayOutputStream) {
+            return new BytesArray(((ByteArrayOutputStream) bos).toByteArray());
+        } else {
+            return ((BytesStream) bos).bytes();
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: Oh boy!

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: A problem for another PR, I assume.

**Review by nik9000 — APPROVED** [review]
Makes sense to me. It is certainly worth looking at the benchmarks after this. I expect if we lose performance we can switch the hot parts to the ctor which takes an `OutputStream` and project `BytesStreamOutput`

File: server/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 1 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: Yep, unfortunately

**Pull request merged**